File: C/B2_main.c
#include <stdio.h>

#define MAX_JOB_QUEUE_SIZE 100
#define cmpAT(a, b) a.arrivalTime < b.arrivalTime
#define cmpBT(a, b) a.burstTime < b.burstTime
#define cmpPRI(a, b) (a.arrivalTime <= b.arrivalTime) && (a.priority > b.priority)

typedef struct Job
{
    int id;
    int arrivalTime;
    int burstTime;
    int priority;
} Job;

struct TableElement
{
    int ID;
    int AT;
    int BT;
    int PRI;
    int CT;
    int TAT;
    int WT;
} op[MAX_JOB_QUEUE_SIZE];

int __jobCnt = 1;

int getNextBatchArrivalTime(Job jobQueue[], int rear, int timestamp)
{
    int res;
    for (res = 0; res < rear && jobQueue[res].arrivalTime <= timestamp; res++)
        ;

    return (res >= rear) ? -1 : jobQueue[res].arrivalTime;
}

Job *getNextShortestJobInQueue(Job jobQueue[], int rear, int timestamp)
{
    int shortest = 0;
    for (int i = 0; i < rear && jobQueue[i].arrivalTime <= timestamp; i++)
    {
        if (jobQueue[shortest].burstTime <= 0)
            shortest = i;
        else if (jobQueue[i].burstTime > 0 && cmpBT(jobQueue[i], jobQueue[shortest]))
            shortest = i;
    }

    return (jobQueue[shortest].burstTime <= 0) ? NULL : &jobQueue[shortest];
}

void initTable(Job jobQueue[], int len)
{
    for (int i = 0; i < len; i++)
        op[jobQueue[i].id] = (struct TableElement){.ID = jobQueue[i].id,
                                                   .AT = jobQueue[i].arrivalTime,
                                                   .BT = jobQueue[i].burstTime,
                                                   .PRI = jobQueue[i].priority};
}

void UpdateTableElementCT(int index, int ct)
{
    op[index].CT = ct;
    op[index].TAT = ct - op[index].AT;
    op[index].WT = (ct - op[index].AT) - op[index].BT;
}

void printTable(int start, int end)
{
    int avgTAT = 0, avgWT = 0;

    printf("\n"
           "\n\t╭──────┬─────┬─────┬──────┬─────┬──────┬─────╮"
           "\n\t│ ID   │ AT  │ BT  │ PRI  │ CT  │ TAT  │ WT  │"
           "\n\t├──────┼─────┼─────┼──────┼─────┼──────┼─────┤");

    for (int i = start; i <= end; i++)
    {
        printf("\n\t│ %-4d │ %-3d │ %-3d │ %-4d │ %-3d │ %-4d │ %-3d │", op[i].ID, op[i].AT, op[i].BT, op[i].PRI,
               op[i].CT, op[i].TAT, op[i].WT);

        avgTAT += op[i].TAT;
        avgWT += op[i].WT;
    }

    printf("\n\t╰──────┴─────┴─────┴──────┴─────┴──────┴─────╯");
    printf("\n\t Avg. TAT: %.2f", (float)avgTAT / (end - start + 1));
    printf("\t   Avg. WT : %.2f", (float)avgWT / (end - start + 1));
}

int __timestamp = 0;

#define sortJobs(jobQueue, len, func)                                                                                  \
    {                                                                                                                  \
        for (int i = 1; i < len; i++)                                                                                  \
            for (int j = i - 1; j >= 0; j--)                                                                           \
                if (func(jobQueue[i], jobQueue[j]))                                                                    \
                {                                                                                                      \
                    Job tmp = jobQueue[i];                                                                             \
                    jobQueue[i] = jobQueue[j];                                                                         \
                    jobQueue[j] = tmp;                                                                                 \
                }                                                                                                      \
    }

void B2_FCFS(Job jobQueue[], int len)
{
    for (int i = 0; i < len; i++)
    {
        if (__timestamp < jobQueue[i].arrivalTime)
        {
            processJob((Job){.id = -1});
            __timestamp = jobQueue[i].arrivalTime;
        }
        processJob(jobQueue[i]);
        __timestamp += jobQueue[i].burstTime;

        UpdateTableElementCT(jobQueue[i].id, __timestamp);
    }
}

void B2_PRI(Job jobQueue[], int len)
{
    sortJobs(jobQueue, len, cmpPRI);
    for (int i = 0; i < len; i++)
    {
        if (__timestamp < jobQueue[i].arrivalTime)
        {
            processJob((Job){.id = -1});
            __timestamp = jobQueue[i].arrivalTime;
        }
        processJob(jobQueue[i]);
        __timestamp += jobQueue[i].burstTime;
        UpdateTableElementCT(jobQueue[i].id, __timestamp);
    }
}

void B2_SRTN(Job jobQueue[], int len)
{
    while (1)
    {
        Job *job = getNextShortestJobInQueue(jobQueue, len, __timestamp);
        if (!job) // All jobs are completed
            break;

        // Job has arrived, process it
        int gap = getNextBatchArrivalTime(jobQueue, len, __timestamp) - __timestamp;
        gap = (gap < 0) || (gap > job->burstTime) ? job->burstTime : gap;
        job->burstTime -= gap;
        processJob(*job);
        __timestamp += gap;
        if (job->burstTime <= 0)
            UpdateTableElementCT(job->id, __timestamp);
    }
}

void B2_RR(Job jobQueue[], int len, int timeSlice)
{
    int i = 0, flag = 0;
    while (flag || i < len)
    {
        // If we have reached the end of the queue AND atleast 1 job is remaining,
        // or if the next job has not arrived yet, reset i to 0
        // reset i to 0
        if ((i >= len && flag) || jobQueue[i].arrivalTime > __timestamp)
        {
            i = 0;
            flag = 0;
        }

        // If job is already completed, skip it
        if (jobQueue[i].burstTime <= 0)
        {
            i++;
            continue;
        }

        // Job has arrived, process it
        int gap = (jobQueue[i].burstTime < timeSlice) ? jobQueue[i].burstTime : timeSlice;
        jobQueue[i].burstTime -= gap;
        processJob(jobQueue[i]);
        __timestamp += gap;
        if (jobQueue[i].burstTime <= 0)
            UpdateTableElementCT(jobQueue[i].id, __timestamp);

        flag = 1;
        i++;
    }
}

void printTimestamp()
{
    printf("\033[42m  %d  \033[44m", __timestamp);
}

void B2_RunAlgos(int i, Job jobQueue[], int len)
{
    initTable(jobQueue, 3);
    __timestamp = 0;
    sortJobs(jobQueue, len, cmpAT);
    printf("\n\033[0m");
    switch (i)
    {
    case 1:
        printf("FCFS:\t");
        B2_FCFS(jobQueue, len);
        break;
    case 2:
        printf("PRI:\t");
        B2_PRI(jobQueue, len);
        break;
    case 3:
        printf("SJF:\t");
        B2_SRTN(jobQueue, len);
        break;
    case 4:
        printf("RR_25:\t");
        B2_RR(jobQueue, len, 25);
        break;
    }
    printTimestamp();
    printTable((i - 1) * len + 1, (i * len));
    printf("\n\n\033[0m");
}

void processJob(Job job)
{
    printTimestamp();
    printf("\033[41m P%d ", job.id);
}

int main()
{
    Job jobQueueFCFS[] = {{.id = __jobCnt++, .priority = 4, .arrivalTime = 0, .burstTime = 100},
                          {.id = __jobCnt++, .priority = 6, .arrivalTime = 0, .burstTime = 90},
                          {.id = __jobCnt++, .priority = 1, .arrivalTime = 2, .burstTime = 5}};

    Job jobQueuePRI[] = {{.id = __jobCnt++, .priority = 4, .arrivalTime = 0, .burstTime = 100},
                         {.id = __jobCnt++, .priority = 6, .arrivalTime = 0, .burstTime = 90},
                         {.id = __jobCnt++, .priority = 1, .arrivalTime = 2, .burstTime = 5}};

    Job jobQueueSRTN[] = {{.id = __jobCnt++, .priority = 4, .arrivalTime = 0, .burstTime = 100},
                          {.id = __jobCnt++, .priority = 6, .arrivalTime = 0, .burstTime = 90},
                          {.id = __jobCnt++, .priority = 1, .arrivalTime = 2, .burstTime = 5}};

    Job jobQueueRR[] = {{.id = __jobCnt++, .priority = 4, .arrivalTime = 0, .burstTime = 100},
                        {.id = __jobCnt++, .priority = 6, .arrivalTime = 0, .burstTime = 90},
                        {.id = __jobCnt++, .priority = 1, .arrivalTime = 2, .burstTime = 5}};

    B2_RunAlgos(1, jobQueueFCFS, 3);

    B2_RunAlgos(2, jobQueuePRI, 3);

    B2_RunAlgos(3, jobQueueSRTN, 3);

    B2_RunAlgos(4, jobQueueRR, 3);

    return 0;
}

File: standalone_impls/Java/B2_main.java
import java.util.*;

class Job {
    int id;
    int arrivalTime;
    int burstTime;
    int priority;

    Job(int id, int arrivalTime, int burstTime) {
        this(id, arrivalTime, burstTime, 0);
    }

    Job(int id, int arrivalTime, int burstTime, int priority) {
        this.id = id;
        this.arrivalTime = arrivalTime;
        this.burstTime = burstTime;
        this.priority = priority;
    }

    void process() {
        AlgoRunner.printTimestamp();
        System.out.printf("\033[41m P%d ", id);
    }

    boolean cmpBT(Job other) {
        return this.burstTime < other.burstTime;
    }
}

class TableElement {
    int ID, AT, BT, PRI, CT, TAT, WT;

    TableElement() {
    }

    TableElement(int id, int at, int bt, int pri) {
        this.ID = id;
        this.AT = at;
        this.BT = bt;
        this.PRI = pri;
    }
}

class Table {
    TableElement[] op = new TableElement[10];

    Table() {
        for (int i = 0; i < 10; i++) {
            op[i] = new TableElement();
        }
    }

    void initTable(List<Job> jobQueue) {
        for (Job job : jobQueue) {
            op[job.id] = new TableElement(job.id, job.arrivalTime, job.burstTime, job.priority);
        }
    }

    void updateCT(int index, int ct) {
        TableElement e = op[index];
        e.CT = ct;
        e.TAT = ct - e.AT;
        e.WT = e.TAT - e.BT;
    }

    void printTable(int start, int end) {
        int avgTAT = 0, avgWT = 0;

        System.out.println("\n\n\t╭──────┬─────┬─────┬──────┬─────┬──────┬─────╮");
        System.out.println("\t│ ID   │ AT  │ BT  │ PRI  │ CT  │ TAT  │ WT  │");
        System.out.println("\t├──────┼─────┼─────┼──────┼─────┼──────┼─────┤");

        for (int i = start; i <= end; i++) {
            TableElement t = op[i];
            System.out.printf("\t│ %-4d │ %-3d │ %-3d │ %-4d │ %-3d │ %-4d │ %-3d │\n",
                    t.ID, t.AT, t.BT, t.PRI, t.CT, t.TAT, t.WT);
            avgTAT += t.TAT;
            avgWT += t.WT;
        }

        System.out.println("\t╰──────┴─────┴─────┴──────┴─────┴──────┴─────╯");
        int count = end - start + 1;
        System.out.printf("\n\t Avg. TAT: %.2f\t   Avg. WT : %.2f\n", (float) avgTAT / count, (float) avgWT / count);
    }
}

class JobQueue {
    List<Job> jobQueue = new ArrayList<>();

    int getNextBatchArrivalTime(int timestamp) {
        for (Job job : jobQueue) {
            if (job.arrivalTime > timestamp)
                return job.arrivalTime;
        }
        return -1;
    }

    Job getNextShortestJobInQueue(int timestamp) {
        Job shortest = null;
        for (Job job : jobQueue) {
            if (job.arrivalTime <= timestamp && job.burstTime > 0) {
                if (shortest == null || job.cmpBT(shortest)) {
                    shortest = job;
                }
            }
        }
        return shortest;
    }
}

class AlgoRunner extends JobQueue {
    private static int __timestamp = 0;
    Table op = new Table();

    public static void printTimestamp() {
        System.out.printf("\033[42m  %d  \033[44m", __timestamp);
    }

    public void B2_RunAlgos(int i) {
        op.initTable(jobQueue);
        __timestamp = 0;
        jobQueue.sort(Comparator.comparingInt(j -> j.arrivalTime));
        System.out.println("\n\033[0m");
        switch (i) {
            case 1:
                System.out.print("FCFS:\t");
                B2_FCFS();
                break;
            case 2:
                System.out.print("PRI:\t");
                B2_PRI();
                break;
            case 3:
                System.out.print("SJF:\t");
                B2_SRTN();
                break;
            case 4:
                System.out.print("RR_25:\t");
                B2_RR(25);
                break;
        }
        printTimestamp();
        op.printTable((i - 1) * jobQueue.size() + 1, i * jobQueue.size());
        System.out.println("\n\n\033[0m");
    }

    private void B2_FCFS() {
        for (Job job : jobQueue) {
            if (__timestamp < job.arrivalTime) {
                new Job(-1, 0, 0).process();
                __timestamp = job.arrivalTime;
            }
            job.process();
            __timestamp += job.burstTime;
            op.updateCT(job.id, __timestamp);
        }
    }

    private void B2_PRI() {
        jobQueue.sort((a, b) -> {
            if (a.arrivalTime != b.arrivalTime)
                return Integer.compare(a.arrivalTime, b.arrivalTime);
            return Integer.compare(b.priority, a.priority);
        });

        for (Job job : jobQueue) {
            if (__timestamp < job.arrivalTime) {
                new Job(-1, 0, 0).process();
                __timestamp = job.arrivalTime;
            }
            job.process();
            __timestamp += job.burstTime;
            op.updateCT(job.id, __timestamp);
        }
    }

    private void B2_SRTN() {
        while (true) {
            Job job = getNextShortestJobInQueue(__timestamp);
            if (job == null)
                break;

            int gap = getNextBatchArrivalTime(__timestamp) - __timestamp;
            if (gap < 0 || gap > job.burstTime)
                gap = job.burstTime;

            job.burstTime -= gap;
            job.process();
            __timestamp += gap;

            if (job.burstTime <= 0)
                op.updateCT(job.id, __timestamp);
        }
    }

    private void B2_RR(int timeSlice) {
        int i = 0;
        boolean flag = false;

        while (flag || i < jobQueue.size()) {
            if ((i >= jobQueue.size() && flag) || jobQueue.get(i).arrivalTime > __timestamp) {
                i = 0;
                flag = false;
            }

            Job job = jobQueue.get(i);
            if (job.burstTime <= 0) {
                i++;
                continue;
            }

            int gap = Math.min(job.burstTime, timeSlice);
            job.burstTime -= gap;
            job.process();
            __timestamp += gap;

            if (job.burstTime <= 0)
                op.updateCT(job.id, __timestamp);

            flag = true;
            i++;
        }
    }
}

public class B2_main {
    public static void main(String[] args) {
        AlgoRunner arFCFS = new AlgoRunner();
        arFCFS.jobQueue.add(new Job(1, 0, 100, 4));
        arFCFS.jobQueue.add(new Job(2, 0, 90, 6));
        arFCFS.jobQueue.add(new Job(3, 2, 5, 1));

        AlgoRunner arPRI = new AlgoRunner();
        arPRI.jobQueue.add(new Job(4, 0, 100, 4));
        arPRI.jobQueue.add(new Job(5, 0, 90, 6));
        arPRI.jobQueue.add(new Job(6, 2, 5, 1));

        AlgoRunner arSRTN = new AlgoRunner();
        arSRTN.jobQueue.add(new Job(7, 0, 100, 4));
        arSRTN.jobQueue.add(new Job(8, 0, 90, 6));
        arSRTN.jobQueue.add(new Job(9, 2, 5, 1));

        AlgoRunner arRR = new AlgoRunner();
        arRR.jobQueue.add(new Job(10, 0, 100, 4));
        arRR.jobQueue.add(new Job(11, 0, 90, 6));
        arRR.jobQueue.add(new Job(12, 2, 5, 1));

        arFCFS.B2_RunAlgos(1);
        arPRI.B2_RunAlgos(2);
        arSRTN.B2_RunAlgos(3);
        arRR.B2_RunAlgos(4);
    }
}
